src/main/java/org/wso2/carbon/connector/connection/EmailConnectionManager.java

```diff
@@ -0,0 +1,163 @@
+        if (connectionMap.get(name) != null) {
+            return connectionMap.get(name);
+        }
+        throw new EmailConnectionException(format("Connection with the name %s has not been initialized.", name));
+    }
+
+    /**
+     * Creates a connection pool
+     *
+     * @param name name of the connection
+     * @param emailConnectionPool email connection pool
+     */
+    private void addConnectionPool(String name, EmailConnectionPool emailConnectionPool) {
+        if (log.isDebugEnabled()){
+            log.debug(format("Creating connection pool for connection : %s", name));
+        }
+        connectionPoolMap.putIfAbsent(name, emailConnectionPool);
+    }
+
+    /**
+     * Retrieves connection pool by name
+     *
+     * @param name name of the connection
+     * @return Email connection Pool
+     * @throws EmailConnectionException if connection pool from the name does not exist
+     */
+    public EmailConnectionPool getConnectionPool(String name) throws EmailConnectionException {
+        if (connectionPoolMap.get(name) != null){
+            if (log.isDebugEnabled()){
+                log.debug(format("Returning connection pool for connection: %s", name));
+            }
+            return connectionPoolMap.get(name);
+        }
+        throw new EmailConnectionException(format("Connection with the name %s has not been initialized.", name));
+    }
+
+    /**
+     * Creates a connection with the given configuration
+     *
+     * @param connectionConfiguration connection configuration
+     */
+    public synchronized void createConnection(ConnectionConfiguration connectionConfiguration) {
+        String connectionName = connectionConfiguration.getConnectionName();
+        if (connectionConfiguration.getProtocol().getName().equalsIgnoreCase(EmailProtocol.SMTP.name())
+                && connectionMap.get(connectionName) == null){
+            // For SMTP protocols a connection pool is not required as they require only a session, which need not be
+            // manipulated as the connection.
+            EmailConnection connection = new EmailConnection(connectionConfiguration);
+            addConnection(connectionConfiguration.getConnectionName(), connection);
+        } else if (!connectionConfiguration.getProtocol().getName().equalsIgnoreCase(EmailProtocol.SMTP.name())
+                && connectionPoolMap.get(connectionName) == null) {
+            // For other protocols, such as IMAP and POP3, connections to a store and folder is made which requires to
+            // handled. Hence, for these instances, we will create a connection pool to optimize the use of these
+            // connections.
+            EmailConnectionPool pool = new EmailConnectionPool(new EmailConnectionFactory(connectionConfiguration),
+                    connectionConfiguration);
+            addConnectionPool(connectionConfiguration.getConnectionName(), pool);
```

Comment:
Are we creating all connections to fill the pool here? Or are we going to create them on demand?

Reply:
They will be created on demand when we invoke `borrowObject`.